### PR TITLE
Adds new algorithm for concave hull calculation - k nearest neighbours approach

### PR DESCRIPTION
**Pull request by Maximkaaa**
The short description of the algorithm is in the docs for the trait.

- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

### CODE REVIEW

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (lnicola) [inline geo/src/algorithm/k_nearest_concave_hull.rs:192]: Would it be possible to use a cross product here (signed area) instead of computing an angle?

**Comment by michaelkirk** [comment]
I was hoping to hold off on reviewing this until this was addressed:

> The first one though looks like an obvious bug. There's actially a check in the code to see if the resulting hull contains all the input points. It probably has to do something with getting points out of the multipolygons. If you send me the geometries, I'll check it out. Any format would be fine. Json is good.

@dabreegster - have you sent the json?

**Comment by dabreegster** [comment]
Sorry for the delay! These're serialized `Multipolygon<f64>`s.

The buggy case:
buggy
[buggy.json.txt](https://github.com/georust/geo/files/6104086/buggy.json.txt)

Some desired "shrinkwrapping" missing:
shrink
[large.json.txt](https://github.com/georust/geo/files/6104091/large.json.txt)

`k=10` for both, but I also tried 3.

**Comment by Maximkaaa** [comment]
Thank you, @dabreegster, your application helped a lot in debugging this algorithm. There were a couple of bugs that I've found and fixed. After fixing, the result in case building concave polygons around the buildings looks like this:
image
image
image

There are still some issues that are not due to implementation but the limitation of the algorithm itself.

1. In some cases polygon looks like it could go more tightly around the objects:

image

The reason for this is that small house in the middle of the park. The algorithm is designed to always build a polygon with single contour. In case of clustered input objects, a rather larger K must be chosen to surround all input objects. One option is to let the method to return multipolygons, but that would break the logic for automatically adjust K, so it would have to be adjusted by hand in many cases.

2. If i switch to surround lanes instead of building, another issue appears:
image
image

The reason for this is that the algorithm in its core only works with points. So to surround other shapes we just take all the vertexes and use them as an input. In the case when there are many little objects (like when we use buildings as input) it's not a problem as all vertexes are close to each other.

In the case when we have big long lines without any points, the algorithm may find some closer points to snap to, so not the whole polygon or line ends up to be inside of the hull. Although all the vertexes are actually inside of the hull.

One way to deal with this issue is to insert virtual vertexes on the long sides of polygons (or lines). But I failed to come up with a good generic way to choose how close I should insert those points.

Another options is to remove this treat from the Lines and Polygons altogether, to make it obvious that the algorithm with deal only with points.

What are your thoughts on this guys?

**Comment by dabreegster** [comment]
> the result in case building concave polygons around the buildings looks like this:

These look much more "shrink-wrapped" -- thanks for iterating on this!

>  In case of clustered input objects, a rather larger K must be chosen to surround all input objects

This limitation makes sense; I don't think manually adjusting K would be feasible for my particular case, so I'm not opinionated about trying it.

> Another options is to remove this treat from the Lines and Polygons altogether, to make it obvious that the algorithm with deal only with points.

This limitation also makes sense, and yes, I would vote for only taking `Point`s in the method signature. I've been a bit confused by other geometric algorithms like generating Voronoi diagrams in the past, because from API documentation, it's not always clear that they don't work on entire polygons, and that you can't just feed in all of the polygon's boundary points.

**Comment by michaelkirk** [comment]
> Another options is to remove this treat from the Lines and Polygons altogether, to make it obvious that the algorithm with deal only with points.

I'd prefer the approach of a signature which only accepts points for now.

To be correct, a hull needs to `Cover` its input. It might be surprising to people that the output could be incorrect in this way if the input is Polygons, especially since other implementations (e.g. PostGIS) do not behave this way. Note that our current concave_hull algorithm also has this limitation, and yet still accepts 1D and 2D geometries, but in hindsight I think this was a mistake.

For followup work, including a recipe in the documentation (and potentially helper methods) for interpolating points along a `line_string` and feeding that into `concave_hull` might be a good middle ground between making things not too tedious while also making clearer the limitations on this algorithm.

Maybe something like:
```
# note: this method doesn't currently exist
let subsampled_linestring = polygon.exterior().interpolated_with_max_distance(0.1);
let points = subsampled_linestring.0;
concave_hull(points)
```

(maybe relevant [`line_interpolate_point`](https://docs.rs/geo/0.18.0/geo/algorithm/line_interpolate_point/trait.LineInterpolatePoint.html))

And also, we should aspire to one day get a fancy concave hull algorithm that also works with 1D and 2D geometries. I'll open a separate issue for this...

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (michaelkirk) [inline geo/src/algorithm/k_nearest_concave_hull.rs:311]: This is admittedly kind of nice, but I'd prefer to avoid adding an extra dependency for it.

(Also, curious if, as @lnicola mentioned, we could remove get_angle altogether)

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (michaelkirk) [inline geo/src/algorithm/k_nearest_concave_hull.rs:18]:
```suggestion
/// 3. As the next point on the hull chose one of the nearest points, that would make the largest
```

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (michaelkirk) [inline geo/src/algorithm/k_nearest_concave_hull.rs:25]:
```suggestion
/// In the worst case scenario, when no K can be found to build a correct hull, the convex hull is
```

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (michaelkirk) [inline geo/src/algorithm/k_nearest_concave_hull.rs:28]:
```suggestion
/// This algorithm is generally several times slower than the one used in the
```

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (michaelkirk) [inline geo/src/algorithm/k_nearest_concave_hull.rs:34]:
```suggestion
/// (lower values do not make sense for this algorithm). If K is equal or larger than the number of
```

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (michaelkirk) [inline geo/src/algorithm/k_nearest_concave_hull.rs:35]:
```suggestion
/// input points, the convex hull will be produced.
```

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (michaelkirk) [reply geo/src/algorithm/k_nearest_concave_hull.rs:139]: Upon further inspection... it seems like this method is only used when the original input has 3 or less points, so orderings here are all equivalent, but that's not obvious by reading this method.

So two alternatives I'd be OK with:
- do an `assert!(points.len() <= 3)` here to document this

or 

- delete this method and bail out earlier in the algorithm when we still have access to the initial point ordering rather than an RTree of those points.

**Review by michaelkirk — CHANGES_REQUESTED** [review]

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (Maximkaaa) [reply geo/src/algorithm/k_nearest_concave_hull.rs:75]: I've changed the logic in points_equal() method and added a comment about this constant.

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (Maximkaaa) [reply geo/src/algorithm/k_nearest_concave_hull.rs:117]: To use the second option we would have to change the algorithm signature to return Option<Polygon<T>>, so I prefer the first option.

File: geo/src/algorithm/k_nearest_concave_hull.rs
Comment (Maximkaaa) [reply geo/src/algorithm/k_nearest_concave_hull.rs]: I've tried using dot-product method, but finally I've found even faster algorithm that works just fine.

**Comment by Maximkaaa** [comment]
Hi, @michaelkirk . Couldn't return to this earlier because of lack of time. I've fixed all the issues, mentioned in the review and updated it to the current master. Please, check if everything fine.

**Comment by michaelkirk** [comment]
Hi @Maximkaaa. I made some time to review and QA the latest version, and things look good to me. Sorry for the slow turn around time!

I'll plan to merge this tomorrow unless I hear from someone else that they'd like to review it.

**Comment by Maximkaaa** [comment]
@michaelkirk NP, totally understand you. It took me a month to make the fixes.

**Comment by michaelkirk** [comment]
Squashed, formatted, and merged in https://github.com/georust/geo/commit/2c4261a4d413e18f3dfbcbc83ad551ae53e775cb

**Pull request closed without merging**
